text/0000-const-generic-const-fn-bounds.md

```diff
@@ -610,26 +616,40 @@ trait Foo {
 which does not force `impl const Foo for Type` to now require passing a `T` with an `impl const Bar`
 to the `a` method.
 
-## `const` function pointers
+## `const` function pointers and `dyn Trait`
+[dynamic_dispatch]: #const-function-pointers-and-dyn-trait
+
+The RFC discusses const bounds on *static* dispatch. What about *dynamic* dispatch?
+In Rust, that means function pointers and `dyn Trait`.
+
+### Function pointers
+
+This is illegal before and with this RFC:
 
 ```rust
 const fn foo(f: fn() -> i32) -> i32 {
     f()
 }
 ```
 
-is illegal before and with this RFC. While we can change the language to allow this feature, two
-questions make themselves known:
+To remain consistent with trait bounds as described in this RFC, it seems
+reasonable to assume that a `fn` pointer passed to a `const fn` would implicitly
+be required to point itself to a `const fn`, and to have an opt-out with
+`?const` for cases where `foo` does not actually want to call `f` (such as
+`RawWakerVTable::new`).
+
+However, there are two problems with this:
 
 1. fn pointers in constants
 
     ```rust
     const F: fn() -> i32 = ...;
     ```
 
-    is already legal in Rust today, even though the `F` doesn't need to be a `const` function.
-    Since we can't reuse this syntax, do we need a different syntax or should we just keep constants
-    as they are and just reuse the syntax in `const fn` arguments?
+    is already legal in Rust today, even though the `F` doesn't need to be a
+    `const` function.  Since we can't reuse this syntax, do we need a different
+    syntax or should the same syntax mean different things for `const` types and
+    `const fn` types?
 
 2. Opt out bounds might seem unintuitive?
```

Comment:
It's not my opinion. Iirc I just added it because someone mentioned it.

Reply:
So what about removing both this and the "natural", basically avoiding making a judgment?

Reply:
I still think it's "natural" or at least consistent to keep `dyn some trait specifiers` in sync with `T: some trait specifiers`

Reply:
I agree with that. And moreover, I think it is just as natural to keep `dyn <specifiers> Trait` in sync with `<specifiers> fn`. Conversely, I think an asymmetry between our two kinds of dynamic dispatch (fn ptrs and `dyn Trait`) is just as unnatural as an asymmetry between traits dispatching statically or dynamically.

So I am fine with keeping the "natural" here for traits if we also call the same thing "natural" for function pointers.

Reply:
but function pointers *already* differ in their specifiers. You can't have `dyn unsafe Trait` for example, but `unsafe fn()` is a valid function pointer type. Not sure if you could even have `extern` fn pointers.

Reply:
`extern` fn ptrs make no sense -- `extern` is a property of a function definition, but irrelevant for its type.

We don't need `unsafe Trait` because the `Trait` declares its methods and whether they are `unsafe`. That is just like saying that function pointers come with argument and return types which `dyn Trait` does not -- it is true, and IMO does not change the argument here. It just reflects that `dyn Trait` is nominal while function pointers are structural.

Imagine a structural version of `dyn Trait`. It would be `dyn trait { fn method1(i32); unsafe fn method2(i64); }`. Does this make the symmetry clearer?

Reply:
yea, ok, the symmetry in semantics is clear, but I was mainly worried about syntax here :laughing:

Reply:
Ah -- I don't care about syntax, as long as it is consistent. ;) Which is what this is not, hence my comments.

Reply:
> So what about removing both this and the "natural", basically avoiding making a judgment?

ok let's do that.

Reply:
Done. I also reordered the fn ptr and dyn Trait discussion, because as you said dyn Trait is closer to what the RFC discusses otherwise, so this makes for a smoother transition I think.